[FSDP2] Zeroed padded tensor in `_apply`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120256
* #119378
* #119302
* __->__ #121509

This PR replaces the `Tensor.resize_` with an explicit zero-ing of the padded tensor. Uninitialized padding is not good since it can give false positive NaNs.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

Commit:
[FSDP2] Zeroed padded tensor in `_apply`

[ghstack-poisoned]

Skylion007 commented on torch/distributed/_composable/fsdp/fully_shard.py:
```diff
@@ -232,12 +232,14 @@ def _apply(self, *args: Any, **kwargs: Any) -> Any:
+                    padded_local_tensor[: local_tensor.size(0)].copy_(local_tensor)
+                    local_tensor = padded_local_tensor
+                fsdp_param._sharded_param_data = local_tensor.view(-1)
+                cast(DTensor, fsdp_param.sharded_param)._local_tensor = local_tensor[
```
Could this cast be handled by an isinstance assert I wonder.

Commit:
Update on "[FSDP2] Zeroed padded tensor in `_apply`"


This PR replaces the `Tensor.resize_` with an explicit zero-ing of the padded tensor. Uninitialized padding is not good since it can give false positive NaNs.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
